### DISCUSSION TITLE
Does terrareg allow us to see services that are using modules?

### DISCUSSION THREAD

**Discussion by sap147** [MatthewJohn/terrareg#29] [Q&A] [answered]
For example, if I publish a module to terrareg, can I see all services consuming that module and the version they are using? All via the terrareg UI?

We are currently exploring moving to a terraform private registry, and were weighing our options on what to use. Thanks.

**Comment by MatthewJohn** [comment, marked as answer]
Hey @sap147 ,

Yes, it does - each service that uses a module should provide an "analytics token", which is used to identify the consumer.

The analytics tokens are provided in the form of a prefix to the namespace in the URL, seperated by double underscore, e.g.:
```
  source  = "registry-demo.mattsbit.co.uk/cloudposse/project-factory/google"
```
Would become
```
  source  = "registry-demo.mattsbit.co.uk/my-project__cloudposse/project-factory/google"
```


When the consumer downloads the module, it will be displayed in the analytics tab of the module:
![Screenshot from 2024-01-03 06-30-18](https://github.com/MatthewJohn/terrareg/assets/1266262/f7fca5de-a8fa-4355-af2c-019b4cc552c7)

By default, Terrareg, analytics are enabled in Terrareg and are enforced, meaning that a consumer _must_ provide an analytics token. Though both the enforcement and the analytics functionality can be disabled.

Let me know if you need any further info and there's also some additional information here: https://github.com/MatthewJohn/terrareg/blob/main/docs/USER_GUIDE.md#module-usage-analytics

Matt

**Reply by sap147** [reply]
Thank you so much. We were basically doing a bake off of Scalr, spacelift, env0, and few other open source solutions. Nothing really made sense for us and most of them were overkill. In terms of the open source offerings, it seems like terrareg is also the only one with a nice UI. We were exploring OpenTofu as well but they dont have a private registry offering. 

Appreciate your help.

**Reply by MatthewJohn** [reply]
Hey, thank you :)

Yes, this is basically exactly the reason Terrareg was made (because the other solutions were a full eco-system and most were "free-mium", as you say.

With regards to OpenTofu, yes, their registry is obviously very new at the moment - without attempting to double guess what they're current plans are - I had the feeling that they were eventually looking at a way to have a local version of their repo, which would replicate the content of their upstream one (I see you created a couple ticket and commented on their RFC ticket about this :) )

During their RFC, one of their maintainers created https://github.com/opentofu/opentofu/issues/811 (referencing here https://github.com/opentofu/opentofu/issues/791#issuecomment-1792288298 and 'Besides' section of https://github.com/opentofu/opentofu/issues/791#issuecomment-1780904505) - so I suspect that they may suggest using alternatives for local registries that don't want to simply be a local replica of theirs :)
If you have any thoughts, maybe adding a comment to https://github.com/opentofu/opentofu/issues/811 might be beneficial

Many thanks
Matt

**Reply by MatthewJohn** [reply]
As a side note, I'm certainly not a UI expert by any means, if there's anything missing or could be improved (though I'm sure there's a lot!) in Terrareg, always feel free to raise any tickets (or PRs! ;) )

**Reply by sap147** [reply]
Thanks! Btw, is there any issue with running/hosting terrareg in EKS? 

Regarding the OpenTofu registry,  I did open [this](https://github.com/opentofu/registry/issues/132#event-11373780535) and so I guess the answer was pretty clear :)
